File: tree/572-subtree-of-another-tree.py
# Solution 1: iterative
# Implementation of solution 2 in C++ file.
# N = number of nodes(S), M = number of nodes(T)
# O(N + M) in space (to store the pre-order array of s and t)
# Time complexity:
# + Step 1: O(N + M)
# + Step 2: O(N * M), N for iterator and M for comparison
# Step 2 can be improved by using pattern matching algorithm
# like KMP --> O(N + M)
class Solution:
    def isSubtree(self, s, t):
        def preorder(node):
            if node is None:
                return [None]
            return [node.val] + preorder(node.left) + preorder(node.right)

        # Step 1: get the pre-order array representation of s and t.
        sp = preorder(s)
        tp = preorder(t)

        # Step 2: check if t is a substring of s
        n = len(tp)
        return any((tp == sp[i:i+n]) for i in range(len(sp)-n+1))

File: tree/572-subtree-of-another-tree.cpp
// Solution 1: recursive
// N = number of nodes(S), M = number of nodes(T)
// O(N + M) in space in worst case (skewed tree)
// O(logN + logM) in space average case
// O(N * M) in time in worst case, but this can be deceiving
// O(N + kM) in time, k = number of occurences of T's root in S.
class Solution {
public:
    bool isSameTree(TreeNode* s, TreeNode* t) {
        if (s == NULL && t == NULL) return true;
        if (s == NULL || t == NULL) return false;
        return s->val == t->val &&
               isSameTree(s->left, t->left) &&
               isSameTree(s->right, t->right);
    }

    bool isSubtree(TreeNode* s, TreeNode* t) {
        if (s == NULL && t == NULL) return true;
        if (s == NULL) return false;
        if (t == NULL) return true;
        return isSameTree(s, t) ||
               isSubtree(s->left, t) ||
               isSubtree(s->right, t);
    }
};


// Solution 2: iterative
// Make use of this statement:
// Two tree has the same structure if they
// have the same pre/post-order traversal
// Please refer to solution 1 in Python file.
